### PR TITLE
Add SQL interaction API

### PR DESCRIPTION
**Pull request by Mats-SX**
This enables the tabular result of a Cypher query to be registered as a named table in the session, and adds a SQL query call to `CAPSSession` which may refer to such registered tables.

### CODE REVIEW

**Comment by Mats-SX** [comment]
Changed my mind; this is ready for review. The rest will come in a follow-up PR.

File: caps-spark/src/main/scala/org/opencypher/caps/impl/spark/CAPSRecords.scala [hunk 1]
```diff
@@ -333,6 +336,16 @@ object CAPSRecords {
...
   }
 
+  /**
+    * Wraps a Spark SQL table (DataFrame) in a CAPSRecords, making it understandable by Cypher.
+    *
+    * @param df the table to wrap.
+    * @param caps the session to which the resulting CAPSRecords is tied.
+    * @return a Cypher table.
+    */
+  def wrap(df: DataFrame)(implicit caps: CAPSSession): CAPSRecords =
```
Comment (s1ck) [inline caps-spark/src/main/scala/org/opencypher/caps/impl/spark/CAPSRecords.scala]: Mh, this re-introduces `CAPSRecords.create(df: DataFrame)` which we removed recently as we don't know really know how to handle the underlying data. Why is this necessary for the SQL functionality? It doesn't seem to be used anywhere except the tests.

File: caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    """|MATCH (p:Person)
+       |RETURN p.age AS age, p.name AS name
+    """.stripMargin
+  )
+
+  // 4) Register the result as a table called people
+  result.records.register("people")
+
+  // 5) Query the registered table using SQL
+  val sqlResults: CypherRecords = session.sql("SELECT age, name FROM people")
```
Comment (s1ck) [inline caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala]: That's nice. Maybe we should add `.sql(query: String)` to `CypherResult` or `CAPSResult`

File: caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+//    drivingTable = Some(sqlRecords))
+
+//  result2.print
+
+}
+
+/**
+  * Specify schema and data with case classes.
+  */
+object SampleData {
```
Comment (s1ck) [inline caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala]: Couldn't you just refer to `Example.SocialNetworkData`?

File: caps-spark/src/test/scala/org/opencypher/caps/impl/spark/CAPSRecordsTest.scala [hunk 4]
```diff
@@ -32,6 +32,45 @@ import scala.collection.Bag
...
+  it("can be registered and queried from SQL") {
+    val givenDF = session.createDataFrame(
+      Seq(
+        (1L, true, "Mats"),
+        (2L, false, "Martin"),
+        (3L, false, "Max"),
+        (4L, false, "Stefan")
+      )).toDF("ID", "IS_SWEDE", "NAME")
+
+    CAPSRecords.wrap(givenDF).register("people")
```
Comment (s1ck) [inline caps-spark/src/test/scala/org/opencypher/caps/impl/spark/CAPSRecordsTest.scala]: you could add `TeamDataFixture` and use the `personDF` or (imho better) the `personTable` which can be instantiated via `CAPSRecords.create(personTable)` and then queried using SQL.

File: caps-spark/src/test/scala/org/opencypher/caps/impl/spark/CAPSSessionImplTest.scala [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import org.opencypher.caps.api.value.CypherValue.CypherMap
+import org.opencypher.caps.test.CAPSTestSuite
+import org.opencypher.caps.test.fixture.TeamDataFixture
+
+import scala.collection.Bag
+
+class CAPSSessionImplTest extends CAPSTestSuite with TeamDataFixture {
+
+  it("can execute sql on registered tables") {
+    CAPSRecords.wrap(personDF).register("people")
```
Comment (s1ck) [inline caps-spark/src/test/scala/org/opencypher/caps/impl/spark/CAPSSessionImplTest.scala]: `CAPSRecords.create(personTable).register("people")`?

**Comment by s1ck** [comment]
Very cool, but I don't think we need the additional `CAPSRecords.wrap` method.

File: caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala (hunk 2 above)
Comment (Mats-SX) [reply caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala]: Good idea.

File: caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala (hunk 3 above)
Comment (Mats-SX) [reply caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala]: Yes probably a cleaner approach. Reason was I added an age field to add some meaning to the query.

**Comment by Mats-SX** [comment]
Changed my mind again, this PR will do on its own.

File: caps-spark/src/main/scala/org/opencypher/caps/impl/spark/CAPSRecords.scala (hunk 1 above)
Comment (Mats-SX) [reply caps-spark/src/main/scala/org/opencypher/caps/impl/spark/CAPSRecords.scala]: I've made it package-private

File: caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala (hunk 2 above)
Comment (Mats-SX) [reply caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala]: We can do that in a while after this is merged.

File: caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala (hunk 3 above)
Comment (Mats-SX) [reply caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala]: Fixed.

File: caps-spark/src/test/scala/org/opencypher/caps/impl/spark/CAPSRecordsTest.scala (hunk 4 above)
Comment (Mats-SX) [reply caps-spark/src/test/scala/org/opencypher/caps/impl/spark/CAPSRecordsTest.scala]: Yes, better idea. Will do that.

**Comment by Mats-SX** [comment]
This depends on the changes made in https://github.com/neo4j/neo4j/pull/10981

**Comment by Mats-SX** [comment]
I've extracted the dependant bits of this PR to #281 to enable this to be merged. #281 seems like it would need to wait for a while.

File: caps-api/src/main/scala/org/opencypher/caps/api/graph/CypherSession.scala [hunk 6]
```diff
@@ -34,7 +35,7 @@ trait CypherSession {
     * @param parameters parameters used by the Cypher query
     * @return result of the query
     */
-  def cypher(query: String, parameters: CypherMap = CypherMap.empty): CypherResult
+  def cypher(query: String, parameters: CypherMap = CypherMap.empty, drivingTable: Option[CypherRecords] = None): CypherResult
```
Comment (s1ck) [inline caps-api/src/main/scala/org/opencypher/caps/api/graph/CypherSession.scala]: As this is the general API for all implementations, I was wondering if that should be only part of CAPSSession (or CORASession later on)

File: caps-api/src/main/scala/org/opencypher/caps/api/graph/PropertyGraph.scala [hunk 7]
```diff
@@ -75,8 +75,8 @@ trait PropertyGraph {
     * @return           the result of the query.
     * @see [[CypherSession.cypherOnGraph()]]
     */
-  final def cypher(query: String, parameters: CypherMap = CypherMap.empty): CypherResult =
-    session.cypherOnGraph(this, query, parameters)
+  final def cypher(query: String, parameters: CypherMap = CypherMap.empty, drivingTable: Option[CypherRecords] = None): CypherResult =
```
Comment (s1ck) [inline caps-api/src/main/scala/org/opencypher/caps/api/graph/PropertyGraph.scala]: see above comment

File: caps-api/src/main/scala/org/opencypher/caps/impl/record/CypherRecords.scala [hunk 8]
```diff
@@ -80,4 +80,11 @@ trait CypherRecords extends CypherTable with CypherPrintable {
     * @return the number of records in this CypherRecords.
     */
   def size: Long
+
+  /**
+    * Registers these records as a table under the given name.
+    *
+    * @param name the name under which this table may be referenced.
+    */
+  def register(name: String): Unit
```
Comment (s1ck) [inline caps-api/src/main/scala/org/opencypher/caps/impl/record/CypherRecords.scala]: see above (should be part of `CORATable` later)

File: caps-spark/src/main/scala/org/opencypher/caps/api/CAPSSession.scala [hunk 9]
```diff
@@ -33,6 +34,8 @@ import scala.reflect.runtime.universe._
 
 trait CAPSSession extends CypherSession {
 
+  def sql(query: String): CypherRecords
```
Comment (s1ck) [inline caps-spark/src/main/scala/org/opencypher/caps/api/CAPSSession.scala]: I think this is the right place.

File: caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  val sqlResults: CypherRecords = session.sql("SELECT age, name FROM people")
+
+//  sqlResults.print
+
+  // 6) Load a purchase network graph via CSV + Schema files
+  val csvFolder = getClass.getResource("/csv/prod/").getFile
+  val purchaseNetwork = session.readFrom(s"file+csv://$csvFolder")
+
+  // 7) Use the results from the SQL query as driving table for a Cypher query
+  val result2 = purchaseNetwork.cypher("WITH name AS name, age AS age MATCH (c:Customer {name: name})-->(p:Product) RETURN c.name, age, p.title", drivingTable = Some(sqlResults))
```
Comment (s1ck) [inline caps-spark/src/test/scala/org/opencypher/caps/demo/CypherSQLRoundtripExample.scala]: very cool

File: caps-spark/src/main/scala/org/opencypher/caps/api/CAPSSession.scala [hunk 11]
```diff
@@ -33,6 +34,8 @@ import scala.reflect.runtime.universe._
 
 trait CAPSSession extends CypherSession {
```
Comment (DarthMax) [inline caps-spark/src/main/scala/org/opencypher/caps/api/CAPSSession.scala]: Maybe add something like 
`trait CAPSSession extends CypherSession with SQLExtensions`

Comment (Mats-SX) [reply caps-spark/src/main/scala/org/opencypher/caps/api/CAPSSession.scala]: Interesting! I'll do that.

File: caps-api/src/main/scala/org/opencypher/caps/api/graph/CypherSession.scala (hunk 6 above)
Comment (Mats-SX) [reply caps-api/src/main/scala/org/opencypher/caps/api/graph/CypherSession.scala]: Well... In Cypher 10 this is going to be the general API; there's a query, optional parameters and an optional driving table.

File: caps-api/src/main/scala/org/opencypher/caps/impl/record/CypherRecords.scala (hunk 8 above)
Comment (Mats-SX) [reply caps-api/src/main/scala/org/opencypher/caps/impl/record/CypherRecords.scala]: I'm not sure... it doesn't seem to far-fetched to consider that this would be supported in Cypher 10... putting it on CORATable would mean we can't use it from the general API without a conversion call hmmm
